### PR TITLE
mullvad-vpn: init at 2019.5

### PR DESCRIPTION
**Pull request by lambdaloop**
###### Motivation for this change

There is no package for the Mullvad VPN client currently. I use this often and perhaps others do as well.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

It mostly fits the CONTRIBUTING document, except that `meta.maintainers` is not set, as I'm not sure what to set it to.

---

### CODE REVIEW

**Comment by eighal** [comment]
Very useful.  Thank you.

File: pkgs/applications/networking/mullvad/default.nix [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    nss
+    udev
+    stdenv.cc.cc
+    systemd
+  ];
+
+in
+
+stdenv.mkDerivation rec {
+  name = "mullvad-${version}";
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
  pname = "mullvad";
```

When `pname` and `version` attributes are present they get concatenated like `name = "${pname}-${version}"` automatically.

File: pkgs/applications/networking/mullvad/default.nix [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    librarypath="${stdenv.lib.makeLibraryPath deps}:$libdir"
+
+    wrapProgram $out/share/mullvad/mullvad-vpn \
+        --prefix LD_LIBRARY_PATH : "$librarypath" \
+        --prefix PATH
+
+    sed -i 's|\/opt\/Mullvad.*VPN|'$out'/bin|g' $out/share/applications/mullvad-vpn.desktop
+    sed -i 's|\/opt\/Mullvad.*VPN/resources|'$out'/bin|g' $out/share/mullvad/resources/mullvad-daemon.conf
+    sed -i 's|\/opt\/Mullvad.*VPN/resources|'$out'/bin|g' $out/share/mullvad/resources/mullvad-daemon.service
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]: Can you install this to `etc/systemd/system/mullvad-daemon.service`.
Then people can use `systemd.packages` in NixOS.

File: pkgs/applications/networking/mullvad/default.nix [hunk 3]
```diff
@@ -0,0 +1,117 @@
+{ stdenv, makeWrapper, fetchurl, dpkg
+, alsaLib, atk, cairo, cups, dbus, expat, fontconfig, freetype
+, gdk_pixbuf, glib, gnome2, nspr, nss, gtk3, gtk2, at-spi2-atk
+, gsettings-desktop-schemas, gobject-introspection, wrapGAppsHook
+, xorg, nghttp2
+, libudev0-shim, glibc, curl, openssl, autoPatchelfHook
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]: I'm seeing arguments here that aren't used, so we should remove them.

File: pkgs/applications/networking/mullvad/default.nix [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    cups
+    dbus
+    expat
+    fontconfig
+    freetype
+    gdk_pixbuf
+    glib
+    gnome2.GConf
+    gnome2.pango
+    gtk2
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mullvad/default.nix [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    cairo
+    cups
+    dbus
+    expat
+    fontconfig
+    freetype
+    gdk_pixbuf
+    glib
+    gnome2.GConf
+    gnome2.pango
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
    pango
```

File: pkgs/applications/networking/mullvad/default.nix [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    gtk2
+    gtk3
+    gsettings-desktop-schemas
+    pulseaudio
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mullvad/default.nix [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    libpulseaudio
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mullvad/default.nix [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    expat
+    fontconfig
+    freetype
+    gdk_pixbuf
+    glib
+    gnome2.GConf
+    gnome2.pango
+    gtk2
+    gtk3
+    gsettings-desktop-schemas
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
```

This has no libraries

File: pkgs/applications/networking/mullvad/default.nix [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    xorg.libXext
+    xorg.libXfixes
+    xorg.libXi
+    xorg.libXrandr
+    xorg.libXrender
+    xorg.libXtst
+    xorg.libxcb
+    nspr
+    nss
+    udev
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
```

File: pkgs/applications/networking/mullvad/default.nix [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    rpath="$out/share/mullvad:$libdir"
+
+    patchelf \
+       --interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+       --set-rpath $rpath $out/share/mullvad/mullvad-vpn
+
+    librarypath="${stdenv.lib.makeLibraryPath deps}:$libdir"
+
+    wrapProgram $out/share/mullvad/mullvad-vpn \
+        --prefix LD_LIBRARY_PATH : "$librarypath" \
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]: I think you can get rid of this completely by doing 
```nix
runtimeDependencies = [ systemd.lib libnotify ];
```

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Hi, thanks for writing the expression @lambdaloop.
The following comments should be clarified or changes should be implemented to reflect them.

Also there might be some confusion on the package name here, in AUR it's [mullvad-vpn](https://repology.org/project/mullvad-vpn/versions) and the actual executable for the GUI is that. So I'm thinking we should go with that name.

**Comment by lambdaloop** [comment]
Hello @worldofpeace, thank you for the comments. I've updated the expression based on your comments. It is much cleaner now and still seems to work fine. Thank you!

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -14347,6 +14347,8 @@ in
 
   morty = callPackage ../servers/web-apps/morty { };
 
+  mullvad = callPackage ../applications/networking/mullvad { };
```
Comment (worldofpeace) [inline pkgs/top-level/all-packages.nix]:
```suggestion
  mullvad-vpn = callPackage ../applications/networking/mullvad-vpn { };
```

You'll need to rename the actual directory as well.

File: pkgs/applications/networking/mullvad/default.nix [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  ];
+
+  buildInputs = deps;
+
+  dontBuild = true;
+  dontConfigure = true;
+
+  unpackPhase = "dpkg-deb -x $src .";
+
+  runtimeDependencies = [systemd.lib libnotify];
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]:
```suggestion
  runtimeDependencies = [ systemd.lib libnotify ];
```

File: pkgs/applications/networking/mullvad/default.nix [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    mv usr/bin/* $out/bin
+    mv opt/Mullvad\ VPN/* $out/share/mullvad
+
+    sed -i 's|\/opt\/Mullvad.*VPN|'$out'/bin|g' $out/share/applications/mullvad-vpn.desktop
+    sed -i 's|\/opt\/Mullvad.*VPN/resources|'$out'/bin|g' $out/share/mullvad/resources/mullvad-daemon.service
+
+    ln -s $out/share/mullvad/mullvad-vpn $out/bin/mullvad-vpn
+    ln -s $out/share/mullvad/resources/mullvad-daemon $out/bin/mullvad-daemon
+
+    install -Dm644 $out/share/mullvad/resources/mullvad-daemon.service $out/etc/systemd/system/mullvad-daemon.service
```
Comment (worldofpeace) [inline pkgs/applications/networking/mullvad/default.nix]: Can you symbolically link this? This duplicates the file.

**Review by worldofpeace — CHANGES_REQUESTED** [review]

**Comment by lambdaloop** [comment]
Hi @worldofpeace , thank you for the fast response. I've updated the package according to your latest comments.

Let me know how it looks.

**Comment by worldofpeace** [comment]
LGTM @lambdaloop. We just need to squash the commits and should be able to merge :sparkles:

**Comment by lambdaloop** [comment]
@worldofpeace squashed!

**Comment by worldofpeace** [comment]
Thanks @lambdaloop :sparkles:

**Comment by eighal** [comment]
Thank you @lambdaloop and @worldofpeace.

**Pull request merged**